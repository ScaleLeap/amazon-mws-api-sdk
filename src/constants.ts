/**
 * The version value is automatically replaced during builds.
 */
export const USER_AGENT = '@scaleleap/amazon-mws-api-sdk/1.9.0'
