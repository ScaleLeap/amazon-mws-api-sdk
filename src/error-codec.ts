import { Codec, exactly, oneOf, optional, string } from 'purify-ts/Codec'

export enum CodeEnum {
  InputStreamDisconnected,
  InvalidParameterValue,
  AccessDenied,
  InvalidAccessKeyId,
  SignatureDoesNotMatch,
  InvalidAddress,
  InternalError,
  'Internal Error',
  QuotaExceeded,
  RequestThrottled,
  ResourceNotFound,
  ScheduledPackageAlreadyExists,
  RegionNotSupported,
  ScheduleWindowExpired,
  InvalidOrderState,
  PickupSlotNotAvailable,
  AccessToFeedProcessingResultDenied,
  ContentMD5Missing,
  ContentMD5DoesNotMatch,
  FeedCanceled,
  FeedProcessingResultNoLongerAvailable,
  FeedProcessingResultNotReady,
  InputDataError,
  InvalidFeedSubmissionId,
  InvalidFeedType,
  InvalidRequest,
  NonRetriableInternalError,
  RetriableInternalError,
  AccessToReportDenied,
  InvalidReportId,
  InvalidReportRequestId,
  InvalidReportType,
  InvalidScheduleFrequency,
  ReportNoLongerAvailable,
  ReportNotReady,
  DependencyFatalException,
  DependencyRetriableException,
  DependencyUnauthorizedException,
  InternalErrorFatalException,
  InvalidInputFatalException,
}

export const Error = Codec.interface({
  Type: string,
  Code: oneOf(Object.keys(CodeEnum).map((x) => exactly(x)) as [Codec<string>, ...Codec<string>[]]),
  Message: string,
  Detail: optional(string),
})

export const MWSApiError = Codec.interface({
  ErrorResponse: Codec.interface({
    Error,
    RequestID: optional(string),
    RequestId: optional(string),
  }),
})
