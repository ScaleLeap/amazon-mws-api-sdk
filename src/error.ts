/* eslint-disable max-classes-per-file */
export class MWSError extends Error {
  constructor(...parameters: string[]) {
    // Propagate some vendor-specific arguments
    super(...parameters)

    // If someone downlevels the compilation target to es5
    Object.setPrototypeOf(this, MWSError.prototype)

    // Maintains proper stack trace (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, MWSError)
    }
  }
}

export class HttpError extends MWSError {
  public message = 'Encountered an error while sending a request'

  constructor(public error: unknown, ...parameters: string[]) {
    super(...parameters)
    Object.setPrototypeOf(this, HttpError.prototype)

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, HttpError)
    }
  }
}

export class ParsingError extends MWSError {
  public message = 'Encountered an error while parsing a response'

  constructor(public error: string, ...parameters: string[]) {
    super(...parameters)
    Object.setPrototypeOf(this, ParsingError.prototype)

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ParsingError)
    }
  }
}
/* eslint-enable max-classes-per-file */
