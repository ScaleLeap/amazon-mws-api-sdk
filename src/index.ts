export * from './http'
export * from './mws'
export * from './sections/sellers'
export * from './sections/orders'
export * from './sections/products/products'
export * from './sections/fulfillment-inventory'
export * from './sections/reports'
export * from './error'
export * from '@scaleleap/amazon-marketplaces'
