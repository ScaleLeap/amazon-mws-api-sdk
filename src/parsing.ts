/** A collection of parsing utils */

export const ensureArray = <T>(x: T | T[]): T[] => (Array.isArray(x) ? x : [x])

export const parseBoolean = (x: 'Yes' | 'No'): boolean => {
  switch (x) {
    case 'Yes':
      return true
    case 'No':
      return false
    default:
      throw new Error('TODO, but should be library-specific error')
  }
}
