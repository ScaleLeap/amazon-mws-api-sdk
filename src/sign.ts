import crypto from 'crypto'

export const sign = (queryString: string, secret: string): string => {
  const hmac = crypto.createHmac('sha256', secret)
  hmac.update(queryString, 'utf8')
  return hmac.digest('base64')
}
