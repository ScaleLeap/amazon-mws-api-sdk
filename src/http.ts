import { AmazonMarketplace } from '@scaleleap/amazon-marketplaces'
import axios from 'axios'
import parser from 'fast-xml-parser'
import { URLSearchParams } from 'url'

import {
  AccessDenied,
  AccessToFeedProcessingResultDenied,
  AccessToReportDenied,
  ContentMD5DoesNotMatch,
  ContentMD5Missing,
  DependencyFatalException,
  DependencyRetriableException,
  DependencyUnauthorizedException,
  enhanceError,
  FeedCanceled,
  FeedProcessingResultNoLongerAvailable,
  FeedProcessingResultNotReady,
  InputDataError,
  InputStreamDisconnected,
  InternalError,
  InternalErrorFatalException,
  InvalidAccessKeyId,
  InvalidAddress,
  InvalidFeedSubmissionId,
  InvalidFeedType,
  InvalidInputFatalException,
  InvalidOrderState,
  InvalidParameterValue,
  InvalidReportId,
  InvalidReportType,
  InvalidRequest,
  InvalidScheduleFrequency,
  InvalidUPCIdentifier,
  NonRetriableInternalError,
  PickupSlotNotAvailable,
  QuotaExceeded,
  RegionNotSupported,
  ReportNoLongerAvailable,
  ReportNotReady,
  RequestThrottled,
  ResourceNotFound,
  RetriableInternalError,
  ScheduledPackageAlreadyExists,
  ScheduleWindowExpired,
  SignatureDoesNotMatch,
} from './error'
import { MWSApiError } from './error-codec'
import { sign } from './sign'

export interface MWSOptions {
  marketplace: AmazonMarketplace
  awsAccessKeyId: string
  mwsAuthToken: string
  sellerId: string
  secretKey: string
}

type HttpMethod = 'GET' | 'POST'
type ParameterTypes = string | number | (number | string)[] | object[] | boolean | undefined
type Parameters = Record<string, ParameterTypes>
type CleanParameters = Record<string, string>

export enum Resource {
  Sellers = 'Sellers',
  Orders = 'Orders',
  Products = 'Products',
  FulfilmentInventory = 'FulfillmentInventory',
  Report = 'Report',
}

interface ResourceActions {
  [Resource.Sellers]:
    | 'ListMarketplaceParticipations'
    | 'ListMarketplaceParticipationsByNextToken'
    | 'GetServiceStatus'
  [Resource.Orders]:
    | 'ListOrders'
    | 'ListOrdersByNextToken'
    | 'GetOrder'
    | 'ListOrderItems'
    | 'ListOrderItemsByNextToken'
    | 'GetServiceStatus'
  [Resource.Products]:
    | 'ListMatchingProducts'
    | 'GetMatchingProduct'
    | 'GetMatchingProductForId'
    | 'GetCompetitivePricingForSKU'
    | 'GetCompetitivePricingForASIN'
    | 'GetLowestOfferListingsForSKU'
    | 'GetLowestOfferListingsForASIN'
    | 'GetLowestPricedOffersForSKU'
    | 'GetLowestPricedOffersForASIN'
    | 'GetMyFeesEstimate'
    | 'GetMyPriceForSKU'
    | 'GetMyPriceForASIN'
    | 'GetProductCategoriesForSKU'
    | 'GetProductCategoriesForASIN'
    | 'GetServiceStatus'
  [Resource.FulfilmentInventory]:
    | 'ListInventorySupply'
    | 'ListInventorySupplyByNextToken'
    | 'GetServiceStatus'
  [Resource.Report]:
    | 'RequestReport'
    | 'GetReportRequestList'
    | 'GetReportRequestListByNextToken'
    | 'GetReportRequestCount'
    | 'CancelReportRequests'
    | 'GetReportList'
    | 'GetReportListByNextToken'
    | 'GetReportCount'
    | 'GetReport'
    | 'ManageReportSchedule'
    | 'GetReportScheduleList'
    | 'GetReportScheduleListByNextToken'
    | 'GetReportScheduleCount'
    | 'UpdateReportAcknowledgements'
}

interface Request {
  url: string
  method: HttpMethod
  headers: Record<string, string>
  data?: string
}

interface ResourceInfo<TResource extends Resource> {
  resource: TResource
  version: string
  action: ResourceActions[TResource]
  parameters: Parameters
}

export interface RequestMeta {
  requestId: string
  timestamp: Date
  quotaMax: number
  quotaRemaining: number
  quotaResetOn: Date
}

interface RequestResponse {
  data: string
  headers: Record<string, string>
}

const canonicalizeParameters = (parameters: CleanParameters): string => {
  const sp = new URLSearchParams(parameters)
  sp.sort()
  return sp.toString().replace(/\+/g, '%20')
}

export const toDotNotation = (object: object, prefix: string) => {
  const result: { [key: string]: string | number | boolean } = {}
  function dotify(plainObject: object, currentKey?: string | number) {
    Object.entries(plainObject).forEach(([key, value]) => {
      const newKey = currentKey ? `${currentKey}.${key}` : key // joined key with dot
      if (value && typeof value === 'object') {
        dotify(value, newKey) // it's a nested object, so do it again
      } else {
        Object.assign(result, { [`${prefix}.${newKey}`]: value }) // it's not an object, so set the property
      }
    })
  }

  dotify(object)
  return result
}

export const cleanParameters = (parameters: Parameters): CleanParameters =>
  Object.entries(parameters)
    .filter(([, parameter]) => parameter !== undefined)
    .reduce((result, [key, parameter]) => {
      if (Array.isArray(parameter)) {
        parameter.forEach((parameterChild: string | number | object, index: number) => {
          if (typeof parameterChild === 'string' || !Number.isNaN(Number(parameterChild))) {
            Object.assign(result, { [`${key}.${index + 1}`]: String(parameterChild) })
          } else {
            Object.assign(result, toDotNotation(parameterChild as object, `${key}.${index + 1}`))
          }
        })
      } else {
        Object.assign(result, { [key]: String(parameter) })
      }

      return result
    }, {} as CleanParameters)

const defaultFetch = ({ url, method, headers, data }: Request): Promise<RequestResponse> =>
  axios({ method, url, headers, data })
    .then((response) => ({
      data: response.data,
      headers: response.headers,
    }))
    .catch((error) => {
      return Promise.reject(error.response.data)
    })

const parseResponse = <T>(response: RequestResponse): [T, RequestMeta] => {
  const responseData = parser.parse(response.data)

  return [
    responseData,
    {
      requestId: response.headers['x-mws-request-id'],
      timestamp: new Date(response.headers['x-mws-timestamp']),
      quotaMax: Number(response.headers['x-mws-quota-max']),
      quotaRemaining: Number(response.headers['x-mws-quota-remaining']),
      quotaResetOn: new Date(response.headers['x-mws-quota-resetson']),
    },
  ]
}

export class HttpClient {
  constructor(
    private options: MWSOptions,
    private fetch: <T>(meta: Request) => Promise<RequestResponse> = defaultFetch,
  ) {}

  public async request<TResource extends Resource, TRes>(
    method: HttpMethod,
    info: ResourceInfo<TResource>,
  ): Promise<[TRes, RequestMeta]> {
    const marketplaceUri = this.options.marketplace.webServiceUri

    const host = marketplaceUri.replace('https://', '')
    const url = `${marketplaceUri}/${info.resource}/${info.version}`

    const parameters = {
      AWSAccessKeyId: this.options.awsAccessKeyId,
      Action: info.action,
      MWSAuthToken: this.options.mwsAuthToken,
      SellerId: this.options.sellerId,
      SignatureMethod: 'HmacSHA256',
      SignatureVersion: '2',
      Timestamp: new Date().toISOString(),
      Version: info.version,
      ...cleanParameters(info.parameters),
    }

    const parametersForSigning = canonicalizeParameters(parameters)
    const queryStringToSign = `${method}\n${host}\n/${info.resource}/${info.version}\n${parametersForSigning}`

    const signature = sign(queryStringToSign, this.options.secretKey)
    const parametersWithSignature = { ...parameters, Signature: signature }

    const headers = {
      'user-agent': '@scaleleap/amazon-mws-api-sdk/1.0.0 (Language=JavaScript)',
    }

    const config =
      method === 'GET'
        ? {
            url: `${url}?${canonicalizeParameters(parametersWithSignature)}`,
            method,
            headers,
          }
        : {
            url,
            method,
            headers,
            data: canonicalizeParameters(parametersWithSignature),
          }

    try {
      const response = await this.fetch(config)

      // GetMatchingProductForId can return an Invalid UPC identifier error message to an otherwise successfully processed request (i.e. 200 status code)
      if (
        info.action === 'GetMatchingProductForId' &&
        response.data.includes('Invalid UPC identifier')
      ) {
        throw new InvalidUPCIdentifier(`${info.action} request failed`)
      }

      return parseResponse(response)
    } catch (error) {
      if (parser.validate(error) !== true) {
        throw error
      }

      const maybeResponse = MWSApiError.decode(parser.parse(error))

      if (maybeResponse.isRight()) {
        const response = maybeResponse.extract()
        const errorCode = response.ErrorResponse.Error.Code

        const errorMap = {
          InputStreamDisconnected,
          InvalidParameterValue,
          AccessDenied,
          InvalidAccessKeyId,
          SignatureDoesNotMatch,
          InvalidAddress,
          InternalError,
          // Subscriptions-specific
          'Internal Error': InternalError,
          QuotaExceeded,
          RequestThrottled,
          ResourceNotFound,
          ScheduledPackageAlreadyExists,
          RegionNotSupported,
          ScheduleWindowExpired,
          InvalidOrderState,
          PickupSlotNotAvailable,
          AccessToFeedProcessingResultDenied,
          ContentMD5Missing,
          ContentMD5DoesNotMatch,
          FeedCanceled,
          FeedProcessingResultNoLongerAvailable,
          FeedProcessingResultNotReady,
          InputDataError,
          InvalidFeedSubmissionId,
          InvalidFeedType,
          InvalidRequest,
          NonRetriableInternalError,
          RetriableInternalError,
          AccessToReportDenied,
          InvalidReportId,
          InvalidReportType,
          InvalidScheduleFrequency,
          ReportNoLongerAvailable,
          ReportNotReady,
          DependencyFatalException,
          DependencyRetriableException,
          DependencyUnauthorizedException,
          InternalErrorFatalException,
          InvalidInputFatalException,
        }

        const ErrorToThrow = errorMap[errorCode]

        throw enhanceError(new ErrorToThrow(`${info.action} request failed`), response)
      } else {
        throw error
      }
    }
  }
}
