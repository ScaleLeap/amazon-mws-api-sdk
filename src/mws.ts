import { HttpClient } from './http'
import { FulfillmentInventory } from './sections/fulfillment-inventory'
import { Orders } from './sections/orders'
import { Products } from './sections/products/products'
import { Reports } from './sections/reports'
import { Sellers } from './sections/sellers'

export class MWS {
  private _sellers!: Sellers

  private _orders!: Orders

  private _fulfillmentInventory!: FulfillmentInventory

  private _products!: Products

  private _reports!: Reports

  constructor(private httpClient: HttpClient) {}

  get sellers() {
    if (!this._sellers) {
      this._sellers = new Sellers(this.httpClient)
    }

    return this._sellers
  }

  get orders() {
    if (!this._orders) {
      this._orders = new Orders(this.httpClient)
    }

    return this._orders
  }

  get fulfillmentInventory() {
    if (!this._fulfillmentInventory) {
      this._fulfillmentInventory = new FulfillmentInventory(this.httpClient)
    }

    return this._fulfillmentInventory
  }

  get products() {
    if (!this._products) {
      this._products = new Products(this.httpClient)
    }

    return this._products
  }

  get reports() {
    if (!this._reports) {
      this._reports = new Reports(this.httpClient)
    }

    return this._reports
  }
}
