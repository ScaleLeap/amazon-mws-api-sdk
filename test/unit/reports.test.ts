import { amazonMarketplaces, HttpClient, ParsingError } from '../../src'
import { MWS } from '../../src/mws'
import { NextToken } from '../../src/parsing'
import { getFixture } from '../utils'

const httpConfig = {
  awsAccessKeyId: '',
  marketplace: amazonMarketplaces.CA,
  mwsAuthToken: '',
  secretKey: '',
  sellerId: '',
}

const headers = {
  'x-mws-request-id': '0',
  'x-mws-timestamp': '2020-05-06T09:22:23.582Z',
  'x-mws-quota-max': '1000',
  'x-mws-quota-remaining': '999',
  'x-mws-quota-resetson': '2020-04-06T10:22:23.582Z',
}

const parsingError = 'Expected an object, but received a string with value ""'

const createMockHttpClient = (fixture: string) =>
  new MWS(
    new HttpClient(httpConfig, () =>
      Promise.resolve({
        data: getFixture(fixture),
        headers,
      }),
    ),
  )

const mockMwsFail = new MWS(
  new HttpClient(httpConfig, () => Promise.resolve({ data: '', headers: {} })),
)

describe('reports', () => {
  describe('getReportListByNextToken', () => {
    const mockNextToken = new NextToken('GetReportList', '123')
    const parameters = {
      NextToken: mockNextToken,
    }

    it('returns report info and next token if succesful', async () => {
      expect.assertions(1)

      const mockGetReportListByNextToken = createMockHttpClient('reports_get_report_list_nt')

      expect(
        await mockGetReportListByNextToken.reports.getReportListByNextToken(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the response isn t valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.reports.getReportListByNextToken(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getReportList', () => {
    const parameters = {}

    it('returns report info and next token if succesful', async () => {
      expect.assertions(1)

      const mockGetReportList = createMockHttpClient('reports_get_report_list')

      expect(await mockGetReportList.reports.getReportList(parameters)).toMatchSnapshot()
    })

    it('throws a parsing error when the response isnt valid', async () => {
      expect.assertions(1)

      await expect(() => mockMwsFail.reports.getReportList(parameters)).rejects.toStrictEqual(
        new ParsingError(parsingError),
      )
    })
  })

  describe('cancelReportRequests', () => {
    const parameters = {}

    it('returns count and detailed info of cancelled requests if succesful', async () => {
      expect.assertions(1)

      const mockCancelReportRequests = createMockHttpClient('reports_cancel_report_requests')

      expect(
        await mockCancelReportRequests.reports.cancelReportRequests(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the response isnt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.reports.cancelReportRequests(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getReportRequestCount', () => {
    const parameters = {}

    it('returns report request count if succesful', async () => {
      expect.assertions(1)

      const mockGetReportRequestCount = createMockHttpClient('reports_get_report_request_count')

      expect(
        await mockGetReportRequestCount.reports.getReportRequestCount(parameters),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the response isn't valid", async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.reports.getReportRequestCount(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getReportRequestListByNextToken', () => {
    const mockNextToken = new NextToken('GetReportRequestList', '123')
    const parameters = {
      NextToken: mockNextToken,
    }

    it('returns report request info if succesful', async () => {
      expect.assertions(1)

      const mockGetReportRequestListByNextToken = createMockHttpClient(
        'reports_get_report_request_list_nt',
      )

      expect(
        await mockGetReportRequestListByNextToken.reports.getReportRequestListByNextToken(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the response isn't valid", async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.reports.getReportRequestListByNextToken(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getReportRequestList', () => {
    const parameters = {}

    it('returns report request info if succesful', async () => {
      expect.assertions(1)

      const mockGetReportRequestList = createMockHttpClient('reports_get_report_request_list')

      expect(
        await mockGetReportRequestList.reports.getReportRequestList(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.reports.getReportRequestList(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('requestReport', () => {
    const parameters = {
      ReportType: '',
    }

    it('returns a parsed model when the response is valid', async () => {
      expect.assertions(1)

      const mockRequestReport = createMockHttpClient('reports_request_report')

      expect(await mockRequestReport.reports.requestReport(parameters)).toMatchSnapshot()
    })

    it('throws a parsing error when the response is not valid', async () => {
      expect.assertions(1)

      await expect(() => mockMwsFail.reports.requestReport(parameters)).rejects.toStrictEqual(
        new ParsingError(parsingError),
      )
    })
  })

  describe('getServiceStatus', () => {
    it('returns a parsed model when the status response is valid', async () => {
      expect.assertions(1)

      const mockMwsServiceStatus = createMockHttpClient('get_service_status')

      expect(await mockMwsServiceStatus.reports.getServiceStatus()).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() => mockMwsFail.reports.getServiceStatus()).rejects.toStrictEqual(
        new ParsingError(parsingError),
      )
    })
  })
})
