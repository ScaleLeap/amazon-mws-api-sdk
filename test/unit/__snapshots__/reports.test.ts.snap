// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports cancelReportRequests returns count and detailed info of cancelled requests if succesful 1`] = `
Array [
  Object {
    "Count": 1,
    "ReportRequestInfo": Object {
      "CompletedDate": undefined,
      "EndDate": 2009-02-13T02:10:39.000Z,
      "GeneratedReportId": undefined,
      "ReportProcessingStatus": "_CANCELLED_",
      "ReportRequestId": "2291326454",
      "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
      "Scheduled": false,
      "StartDate": 2009-01-21T02:10:39.000Z,
      "StartedProcessingDate": undefined,
      "SubmittedDate": 2009-02-20T02:10:39.000Z,
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getReportList returns report info and next token if succesful 1`] = `
Array [
  Object {
    "HasNext": true,
    "NextToken": NextToken {
      "action": "GetReportList",
      "token": "2YgYW55IPQhvu5hbCBwbGVhc3VyZS4=",
    },
    "ReportInfo": Object {
      "Acknowledged": false,
      "AcknowledgedDate": undefined,
      "AvailableDate": 2009-02-10T09:22:33.000Z,
      "ReportId": "898899473",
      "ReportRequestId": "2278662938",
      "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getReportListByNextToken returns report info and next token if succesful 1`] = `
Array [
  Object {
    "HasNext": false,
    "NextToken": NextToken {
      "action": "GetReportList",
      "token": "none",
    },
    "ReportInfo": Object {
      "Acknowledged": false,
      "AcknowledgedDate": undefined,
      "AvailableDate": 2009-02-10T09:22:33.000Z,
      "ReportId": "898899473",
      "ReportRequestId": "2278662938",
      "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getReportRequestCount returns report request count if succesful 1`] = `
Array [
  Object {
    "Count": 1276,
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getReportRequestList returns report request info if succesful 1`] = `
Array [
  Object {
    "HasNext": true,
    "NextToken": NextToken {
      "action": "GetReportRequestList",
      "token": "2YgYW55IPQhcm5hbCBwbGVhc3VyZS4=",
    },
    "ReportRequestInfo": Object {
      "CompletedDate": 2011-02-17T23:44:48.000Z,
      "EndDate": 2011-02-13T02:10:39.000Z,
      "GeneratedReportId": "3538561173",
      "ReportProcessingStatus": "_DONE_",
      "ReportRequestId": "2291326454",
      "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
      "Scheduled": false,
      "StartDate": 2011-01-21T02:10:39.000Z,
      "StartedProcessingDate": 2011-02-17T23:44:43.000Z,
      "SubmittedDate": 2011-02-17T23:44:09.000Z,
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getReportRequestListByNextToken returns report request info if succesful 1`] = `
Array [
  Object {
    "HasNext": false,
    "NextToken": NextToken {
      "action": "GetReportRequestList",
      "token": "none",
    },
    "ReportRequestInfo": Object {
      "CompletedDate": undefined,
      "EndDate": 2009-02-13T02:10:39.000Z,
      "GeneratedReportId": undefined,
      "ReportProcessingStatus": "_SUBMITTED_",
      "ReportRequestId": "2291326454",
      "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
      "Scheduled": false,
      "StartDate": 2009-01-21T02:10:39.000Z,
      "StartedProcessingDate": undefined,
      "SubmittedDate": 2009-02-20T02:10:39.000Z,
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports getServiceStatus returns a parsed model when the status response is valid 1`] = `
Array [
  Object {
    "Status": "GREEN",
    "Timestamp": "2020-05-06T08:22:23.582Z",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`reports requestReport returns a parsed model when the response is valid 1`] = `
Array [
  Object {
    "CompletedDate": undefined,
    "EndDate": 2009-02-13T02:10:39.000Z,
    "GeneratedReportId": undefined,
    "ReportProcessingStatus": "_SUBMITTED_",
    "ReportRequestId": "2291326454",
    "ReportType": "_GET_MERCHANT_LISTINGS_DATA_",
    "Scheduled": false,
    "StartDate": 2009-01-21T02:10:39.000Z,
    "StartedProcessingDate": undefined,
    "SubmittedDate": 2009-02-20T02:10:39.000Z,
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;
